src/Elastic.Apm/CurrentExecutionSegmentsContainer.cs

```diff
@@ -3,22 +3,22 @@
 // See the LICENSE file in the project root for more information
 
 using System.Threading;
-using Elastic.Apm.Model;
+using Elastic.Apm.Api;
 
 namespace Elastic.Apm
 {
 	internal sealed class CurrentExecutionSegmentsContainer : ICurrentExecutionSegmentsContainer
 	{
-		private readonly AsyncLocal<Span> _currentSpan = new AsyncLocal<Span>();
-		private readonly AsyncLocal<Transaction> _currentTransaction = new AsyncLocal<Transaction>();
+		private readonly AsyncLocal<ISpan> _currentSpan = new AsyncLocal<ISpan>();
+		private readonly AsyncLocal<ITransaction> _currentTransaction = new AsyncLocal<ITransaction>();
 
-		public Span CurrentSpan
+		public ISpan CurrentSpan
 		{
 			get => _currentSpan.Value;
 			set => _currentSpan.Value = value;
 		}
 
-		public Transaction CurrentTransaction
+		public ITransaction CurrentTransaction
 		{
 			get => _currentTransaction.Value;
```

Comment:
Here we still return `null` if there is no current transaction (same applies to `CurrentSpan`) and this causes a bit of a pain on the current Public API.

Writing code like this isn't safe:

```
agent.Tracer.CurrentTransaction.CaptureSpan("foo", "bar", () =>
{
	//user code to run
});
```

`agent.Tracer.CurrentTransaction` can throw a nullref if the current transaction is null, even worse, accidentally people may use `?.`:

```
agent.Tracer.CurrentTransaction?.CaptureSpan("foo", "bar", () =>
{
	//user code to run
});
```

But this cause the bigger problem that the user code in the lambda won't run if the current transaction is null.

This problem could be fixed if we'd return `NoopTransaction` here. The problem with doing that is that I feel it'd be a major breaking change: right now we force people to null-check here, if we change this to never return `null` then we could cause issues in user code. Same applies to `HttpContextCurrentExecutionSegmentsContainer.cs`.

Therefore I decided that in this PR I don't change this behaviour. Everything is already in place to do this change at some point, but I think best would be to do this on a major version bump.

I'd be happy to hear if this makes sense or I should reconsider.

Reply:
> The problem with doing that is that I feel it'd be a major breaking change: right now we force people to null-check here, if we change this to never return null then we could cause issues in user code. Same applies to HttpContextCurrentExecutionSegmentsContainer.cs.

I'm guessing that folks end up writing code in a way that always runs their code, irrespective of whether there is a transaction or not i.e. if there's a transaction, run code inside it, if not, run anyway. Is there a particular pattern that you foresee, where returning a `NoopTransaction` would cause an issue? I could see one if the action taken is different depending on if there is a transaction or not.

I _think_ this behavioural change feels OK to go into a minor version. There's a chance it could cause issues in some existing user code, but it'd greatly simplify agent usage

Reply:
Ok - agreed. But I'd do it outside this PR. I opened https://github.com/elastic/apm-agent-dotnet/issues/1017.